nixos/modules/system/boot/luksroot.nix

```diff
@@ -853,6 +876,31 @@ in
                       -> versionAtLeast kernelPackages.kernel.version "5.9";
           message = "boot.initrd.luks.devices.<name>.bypassWorkqueues is not supported for kernels older than 5.9";
         }
+
+        { assertion = config.boot.initrd.systemd.enable -> all (dev: !dev.fallbackToPassword) (attrValues luks.devices);
+          message = "boot.initrd.luks.devices.<name>.fallbackToPassword is implied by systemd stage 1.";
+        }
+        { assertion = config.boot.initrd.systemd.enable -> all (dev: dev.preLVM) (attrValues luks.devices);
+          message = "boot.initrd.luks.devices.<name>.preLVM is not used by systemd stage 1.";
+        }
+        { assertion = config.boot.initrd.systemd.enable -> options.boot.initrd.luks.reusePassphrases.highestPrio == defaultPrio;
+          message = "boot.initrd.luks.reusePassphrases has no effect with systemd stage 1.";
+        }
```

Comment:
Should these three be [warnings](https://nixos.org/manual/nixos/stable/index.html#sec-assertions-warnings)?

Reply:
No, why would they be?

Reply:
I imagine the rationale is that these don't _break_ anything if set when systemd stage 1 is enabled, and therefore they should be warnings as opposed to assertions.

I'm uncertain whether that is correct though, since setting something that should affect boot but doesn't might break your boot, even if it didn't break the config eval :)

Reply:
FWIW, having `preLVM` or `reusePassphrases` set won't break your expectations, as systemd stage 1 does the correct thing either way so it will do what you wanted without regard for those options. Disabling `fallbackToPassword` would have unexpected results if we didn't error the eval though.

I don't particularly care one way or the other about warnings vs assertions for `preLVM` and `reusePassphrases` though

Reply:
Yeah, that's exactly the motivation. If you set something that is not processed by the new initrd, it might prevent your system from booting.

Reply:
My suggestion was that *these particular* options (`fallbackToPassword`, `preLVM`, `reusePassphrases`) will not prevent your system from booting, so they could be downgraded to warnings, in order to make things a bit easier for someone who wants to flip between the traditional and systemd initrds for testing.

Reply:
Disabling `fallbackToPassword` definitely *can* cause your system to fail to boot if the option is just ignored. You expect it to fallback to your password but it doesn't and it fails instead.

Reply:
@ElvishJerricco The message says “fallbackToPassword is implied by systemd stage 1”—i.e. it is always enabled, not always disabled.

Reply:
@andersk That's not exactly what's meant. Enabling systemd-stage1 doesn't force the fallbackToPassword NixOS option to be enabled; the PR is currently just informing the user that the behavior of systemd-stage1 is that of fallbackToPassword, and we don't have a way to disable it. i.e. we *cannot* set it to false, so if you do that, we need to make sure you know that will not do what you expect; we just ignore the option and fallback in stage 1 anyway.

Reply:
I understand that doesn’t literally alter the option to evaluate to `true`, it just has that behavior. But I don’t see how to reconcile “You expect it to fallback to your password but it doesn't and it fails instead” with “the behavior of systemd-stage1 is that of fallbackToPassword”. If you expect it to fall back to your password, and it always falls back to your password, then it has done what you expect. Why should it prevent you from configuring it to do what it going to do anyway?

Reply:
@andersk I'm suggesting that if a user writes `fallbackToPassword = false;` in their nixos configuration, we need that to be an error, not a warning, because it will not behave as expected.

Reply:
"need" was too strong a word. I don't think we really need to worry about this concern very much in the first place TBH.

Reply:
> @andersk I'm suggesting that if a user writes `fallbackToPassword = false;` in their nixos configuration, we need that to be an error, not a warning, because it will not behave as expected.

That would be a reasonable check. It is not, however, the check performed by this code. This code throws an error on `fallbackToPassword = true`.

Reply:
What if people did `fallbackToPassword = lib.mkIf (!config.boot.initrd.systemd.enable) false`?

Reply:
Keeping this an assertion makes sure there are no issues. Making it a warning might cause issues.

Let's please land this and revisit before making the options public. Assertion failures when switching stage-1 implementations are to be expected, especially at this stage.

We do not need to tackle the distinction between what should be a warning or an assertion right now. 

Let's first make it work, then we can worry about making it good, and finally, we can make it easy :)

Reply:
https://github.com/NixOS/nixpkgs/projects/51#card-81007876